pkg/openwsn/sock/openwsn_sock_udp.c

```diff
@@ -0,0 +1,549 @@
+        src_addr.type = ADDR_128B;
+        memcpy(&src_addr.addr_128b, &sock->gen_sock.local.addr,
+               LENGTH_ADDR128b);
+    }
+
+    /* get pkt buffer and take owner ship */
+    if ((pkt = openqueue_getFreePacketBuffer(COMPONENT_SOCK_TO_UDP)) == NULL) {
+        return -ENOMEM;
+    }
+    pkt->owner = COMPONENT_SOCK_TO_UDP;
+    pkt->creator = COMPONENT_SOCK_TO_UDP;
+
+    /* set addresses and ports */
+    pkt->l3_destinationAdd.type = ADDR_128B;
+    pkt->l3_sourceAdd.type = ADDR_128B;
+    memcpy(&pkt->l3_sourceAdd.addr_128b, &src_addr.addr_128b, LENGTH_ADDR128b);
+    memcpy(&pkt->l3_destinationAdd.addr_128b, &dst_addr.addr_128b,
+           LENGTH_ADDR128b);
+    pkt->l4_destination_port = dst_port;
+    pkt->l4_sourcePortORicmpv6Type = src_port;
+
+    /* set payload */
+    if (packetfunctions_reserveHeader(&pkt, len)) {
+        openqueue_freePacketBuffer(pkt);
+        return -ENOMEM;
+    }
+    memcpy(pkt->payload, data, len);
+    pkt->l4_payload = pkt->payload;
+    pkt->l4_length = pkt->length;
+
+    /* push task to scheduler send */
+    scheduler_push_task(_sock_transmit_internal, TASKPRIO_UDP);
+
+    return len;
+}
+
+void sock_udp_close(sock_udp_t *sock)
+{
+    if (_udp_socket_list == NULL) {
+        return;
+    }
+    if (sock) {
+        LL_DELETE(_udp_socket_list, sock);
+        sock->next = NULL;
+    }
+}
+
+int sock_udp_get_local(sock_udp_t *sock, sock_udp_ep_t *ep)
+{
+    if (sock->gen_sock.local.family == AF_UNSPEC) {
+        return -EADDRNOTAVAIL;
+    }
+
+    memcpy(ep, &sock->gen_sock.local, sizeof(sock_udp_ep_t));
+
+    return 0;
+}
+
+int sock_udp_get_remote(sock_udp_t *sock, sock_udp_ep_t *ep)
+{
+    if (sock->gen_sock.remote.family == AF_UNSPEC) {
+        return -ENOTCONN;
+    }
+
+    memcpy(ep, &sock->gen_sock.remote, sizeof(sock_udp_ep_t));
+
+    return 0;
+}
+
+ssize_t sock_udp_recv(sock_udp_t *sock, void *data, size_t max_len,
+                      uint32_t timeout, sock_udp_ep_t *remote)
+{
+    void *pkt = NULL, *ctx = NULL;
+    ssize_t tmp, res;
+
+    assert((sock != NULL) && (data != NULL) && (max_len > 0));
+    res = sock_udp_recv_buf(sock, &pkt, &ctx, timeout, remote);
+    /* set data to copy, check if enough buffer space */
+    if (res >= 0) {
+        if ((ssize_t)max_len >= res) {
+            memset(data, 0, max_len);
+            if (res > 0) {
+                /* copy received data */
+                memcpy(data, pkt, res);
+                /* free packet */
+                tmp = sock_udp_recv_buf(sock, &pkt, &ctx, timeout, remote);
+                assert(tmp == 0);
+            }
+        }
+        else {
+            res = -ENOBUFS;
+            /* free packet */
+            tmp = sock_udp_recv_buf(sock, &pkt, &ctx, timeout, remote);
+            assert(tmp == 0);
+        }
+    }
+    return res;
+}
+
+ssize_t sock_udp_recv_buf(sock_udp_t *sock, void **data, void **buf_ctx,
+                          uint32_t timeout, sock_udp_ep_t *remote)
```

Comment:
Can I haz test back for this function back plz

![](https://media.giphy.com/media/I1nwVpCaB4k36/giphy.gif)

Reply:
I had added the same that is in `gnrc_sock_udp` https://github.com/RIOT-OS/RIOT/blob/3ed138242bf76195cc9ccdfb417721060c8f80e2/tests/pkg_openwsn_sock_udp/main.c#L432-L455

Reply:
Which commit?

Reply:
Ah 02d3a601773... somehow missed that